### ISSUE TITLE
document.cookie doesn't get set on the page loaded via the custom protocol

### ISSUE THREAD

**Issue by vladimiry** [electron/electron#27981]
* **Electron Version:** 12.0.0 (actually any version, I tested many except from the beta/nightly)
* **Operating System:** Linux
* **Last Known Working Electron version:** none

### Expected Behavior
The `document.cookie:  is-paid-user=1` gets printed to console.

### Actual Behavior
The `document.cookie: ` gets printed to console (no `is-paid-user=1` cookie record).

### To Reproduce
After changing the `web://news.ycombinator.com` to https://news.ycombinator.com the issue will be gone.

```javascript
const {app, BrowserWindow, protocol} = require("electron");

const scheme = "web";

protocol.registerSchemesAsPrivileged([{scheme, privileges: {secure: true, standard: true, corsEnabled: true}}]);

app.whenReady().then(() => {
  const window = new BrowserWindow();

  window.webContents.session.protocol.registerBufferProtocol(scheme, (...[, callback]) => {
    callback({mimeType: "text/html", data: Buffer.from("<h5>blank page</h5>")})
  })

  window.loadURL("web://news.ycombinator.com").then(() => {
    window.webContents.openDevTools();
    window.webContents.executeJavaScript(`
      document.cookie = "is-paid-user=1;expires=Fri, 31 Dec 2021 20:59:59 GMT;domain=.ycombinator.com;path=/";
      console.log("document.cookie: ", document.cookie);
    `);
  });
});
```

### Screenshots

- https scheme:
  ![expected](https://user-images.githubusercontent.com/1560781/109812794-90419b00-7c3d-11eb-9b76-5f3d97e1446d.png)
- web/custom scheme:
  ![actual](https://user-images.githubusercontent.com/1560781/109812809-946db880-7c3d-11eb-98da-f579daa01c60.png)

### Additional Information

My understanding was that setting `standard: true` in `registerSchemesAsPrivileged` call is enough to make `document.cookie` work on the custom schemes.

**Comment by famingyuan** [comment]
Is there any solutions?

**Comment by nornagon** [comment]
It's failing on this line: https://source.chromium.org/chromium/chromium/src/+/master:net/cookies/cookie_util.cc;l=268;drc=587068ee20325543d40cbd4bf50bed6af51ee23e

because 

```
cookie_domain_and_registry = ycombinator.com
url_domain_and_registry = news.ycombinator.com
```

changing the cookie to be for `news.ycombinator.com` results in [this check](https://source.chromium.org/chromium/chromium/src/+/master:net/cookies/cookie_monster.cc;l=1767;drc=587068ee20325543d40cbd4bf50bed6af51ee23e) being hit instead, because the cookieable schemes don't include `web`. Unfortunately it looks like the only way to affect that list without patching chrome is through [CookieManagerParams.allow_file_scheme_cookies](https://source.chromium.org/chromium/chromium/src/+/master:services/network/public/mojom/cookie_manager.mojom;l=30;drc=587068ee20325543d40cbd4bf50bed6af51ee23e), which would add `file:` to the cookieable scheme list. There's no way to add arbitrary schemes to it.

tl;dr we don't support cookies on custom protocols, and can't without a fairly significant effort.


... and it looks like i previously did this same investigation here: https://github.com/electron/electron/issues/20940#issuecomment-565578759

I'm changing this to a feature request. If someone wants to tackle it, I'd be happy to review PRs, but fair warning: I'll ask you to upstream any patches to Chromium.

**Comment by vladimiry** [comment]
> tl;dr we don't support cookies on custom protocols ...

Is this documented somewhere at https://www.electronjs.org/docs?

**Comment by nornagon** [comment]
@vladimiry I don't think so... I'm not sure where would be the best place to document. Do you want to make a docs PR?

**Comment by vladimiry** [comment]
> Do you want to make a docs PR?

Not ready to jump into it.

But would be great the docs get updated since I've reviewed it before placing the issue and the following part mislead me: https://github.com/electron/electron/blame/80f89a34728281cc61cfacd7b682f0227f89a102/docs/api/protocol.md#L101-L104 (notices `cookies` mentioning)

> By default web storage apis (localStorage, sessionStorage, webSQL, indexedDB,
cookies) are disabled for non standard schemes. So in general if you want to
register a custom protocol to replace the `http` protocol, you have to register
it as a standard scheme.

> I'm not sure where would be the best place to document.

So maybe the adjustment would land somewhere close to this paragraph.

**Comment by doug-cb** [comment]
> Unfortunately it looks like the only way to affect that list without patching chrome is through [CookieManagerParams.allow_file_scheme_cookies](https://source.chromium.org/chromium/chromium/src/+/master:services/network/public/mojom/cookie_manager.mojom;l=30;drc=587068ee20325543d40cbd4bf50bed6af51ee23e), which would add `file:` to the cookieable scheme list.

Is this currently possible?

I'm running into an issue using a third-party SDK that relies upon cookies to work. It works just fine in development since I'm using a local dev server, but I haven't gotten it to work in a packaged state via either file: or a custom protocol.

The only other option I see is to run a local server, even in production, and server up the file via http, which would be less than ideal.

**Comment by giuliopaci** [comment]
Hi everybody! Any news on this issue?
As most of the other people I also got confused, mostly, by: https://github.com/electron/electron/blame/80f89a34728281cc61cfacd7b682f0227f89a102/docs/api/protocol.md#L101-L104

I think the documentation for custom protocols should clearly report that cookies are not supported.
Ideally somewhere there should be some samples about how to setup some common scenarios.

For example, in my case I am trying to package a single page application written using react that access a remote backend. The backend has proper configuration for CORS and CSRF (expecting the client to set a header with the content read from a cookie).

Following some hints that I found online, I was able to tweak CORS backend settings in the application, using `webRequest.onHeadersReceived` and `webRequest.onBeforeSendHeaders`. So all GET requests worked properly using `file://` protocol. After some investigation, from the documentation I found that adding a custom protocol and registering as standard it would have allowed me to replace `http` protocol completely... But after several attempts I realized that it is not possible to set cookies. Even from the console, running
```
window.cookieStore.set('key', 'value');
```
resulted in:
```
Uncaught (in promise) DOMException: An unknown error occurred while writing the cookie.
```
and thus decided to search for a bug (I think there is a bug, since the documentation clearly states that a custom protocol registered as standard can serve the purpose to replace http protocol).

Is there any easy approach to solve my use case without setting up a local server?

**Comment by zamitto** [comment]
Any updates or workaround on this?
I'm also running a third party sdk which uses cookies

**Comment by mmarczell-graphisoft** [comment]
@zamitto 

For what it's worth, the third party SDK I am using (MS Application Insights) exposes a way to redirect its cookie storage requirements to another storage system such as localStorage.

@giuliopaci 

> Instead of defining my own custom scheme, I have tried to intercept the http protocol `protocol.interceptFileProtocol('http', handler)` and it seems to work, using exactly the same handler that I previously used with `protocol.registerFileProtocol('app', handler)`. Is there any potential issue with this approach?

The Electron Security guide recommends using a custom protocol

https://www.electronjs.org/docs/latest/tutorial/security#18-avoid-usage-of-the-file-protocol-and-prefer-usage-of-custom-protocols
